Populate profiling events with source info.

kimishpatel opened the pull request:
Summary:
This changelist adds:
1) Information about where the operators are in the model source.
2) Populates this information in chrome trace.
3) Enables use of python interface for profiler to get source debug info.

Differential Revision: D15772906

soumith commented:
i think this introduces overhead into what is a high-performance profiler.
Please make sure you benchmark worst-case profiling, and have a review by @zdevito @apaszke @dzhulgakov

kimishpatel commented:
@soumith thanks. will certainly benchmark it, but is there a specific code pattern you saw that worried you? I have tried to kept the overhead to minimum such that during graph execution we only update thread local variables (two 64 bit values).

soumith commented:
I hadn't fully gone through the PR previously. I just did, and I see that it is **only** for graph mode (no eager support). Is that deliberate?

kimishpatel commented:
@soumith, at the moment yes. Although if landing this requires support for both eager mode and graph mode then I will update the PR appropriately.

soumith commented:
adding eager mode support will make the profiler very slow (my original comment assumed you did for eager mode as well), because inspecting python source or getting stack info is not cheap

kimishpatel commented:
I see. Had not looked into that. Thanks for that comment.

dzhulgakov commented on torch/csrc/jit/interpreter.cpp:
```diff
@@ -434,6 +444,13 @@ struct CodeImpl {
   void insertInstruction(OpCode op, int64_t X = 0, uint64_t N = 0) {
     instructions_.emplace_back(op, X, N);
     instructions_source_.emplace_back(current_node_);
+    if (torch::autograd::profiler::hasCallbacks()) {
```
the problem is that insertInstruction gets called only once when the module is first executed, and not on each forward(). So it's possible that profiler was initialized only after.

I also feel that if profiling is on, storing shared_ptr<CodeImpl> (or some separate debug data struct) might be preferable than proactively stringifying all instructions.

kimishpatel replied:
@dzhulgakov, regarding when the profiler is initialized comment: We can record debug information about graph nodes unconditionally. This still keeps events logging the information needed conditional.

Regarding your second point: sure. We can store shared_ptr to Node perhaps.

dzhulgakov commented on torch/csrc/jit/interpreter.cpp:
```diff
@@ -434,6 +444,13 @@ struct CodeImpl {
+      debug_info_str = current_node_->sourceRange().text();
+      replaceAllChar('\n', debug_info_str);
+      int64_t graph_id = graph_->getGraphId();
+      torch::autograd::profiler::setSourceDebugInfo(graph_id, debug_info_str);
```
btw, the contract of appending the instructions in order is pretty fragile

kimishpatel replied:
@dzhulgakov, you are right. I followed onto what is already assumed. I still do agree with you. If the implementation of how instructions are stored changes it requires corresponding change here too. Let me think how to address this better.

dzhulgakov commented on torch/csrc/jit/ir.h:
```diff
@@ -972,6 +973,11 @@ struct Graph {
+  // This is used to identify instance of a graph uniquely within a process.
+  // This is used for source range augmentation in profiling that generates
+  // a trace.
+  int64_t graph_id_{-1};
```
it feels like the wrong place to tie things together. As commented above, I'd rather keep some pointer to CodeImpl or Graph itself inside profiler (we can downcast it to shared_ptr<void> if needed) instead of polluting JIT code

kimishpatel replied:
@dzhulgakov, so every event needs two pieces of information. 1) which code and 2) what pc. Earlier I did 1) by storing graph_id_, but this is wrong, so instead I need something equivalent of code_id. Trouble with storing shared_ptr<CodeImpl> for every event is that you will need to store shared_ptr which will involve atomic counters and this is what I wanted to avoid for per reason. I think it is possible to keep all the code inside interpreter.cpp.

1. Each CodeImpl instance has a unique id and source_debug_info maintains a map that maps from this unique id to shared_ptr<CodeImpl>. This is done each time Code is generated, likely at the creation of ExecutionPlan.
2. During the execution then each even is annotated by CodeImpl instance's unique id and the pc.

Note I still feel stringifying debug info at the time of instruction insertion is better than doing at the time of trace creation since we will be creating and destroying lot of string otherwise.

kimishpatel replied:
Also I realize CodeImpl or Code is part of interpreter, so storing a shared pointer to it requires including `interpreter.h`. This I feel is not appropriate as profiling does not need access to interpreter.h.

dzhulgakov commented on torch/csrc/autograd/source_debug_info.cpp:
```diff
@@ -0,0 +1,45 @@
+#include  <torch/csrc/autograd/source_debug_info.h>
+
+namespace torch { namespace autograd { namespace profiler {
+std::mutex module_source_debug_info_lock;
```
global locks are evil, especially in multithreaded environment (e.g. in predictor). Can we instead attach debug info at Function level or rely on shared_ptrs?

kimishpatel replied:
@dzhulgakov, this lock is used only at the time of code creation. Is it still a concern?

jamesr66a replied:
Code creation is called under a compilation mutex in GraphExecutor: https://github.com/pytorch/pytorch/blob/master/torch/csrc/jit/graph_executor.cpp#L532

Can we rely on that instead?

dzhulgakov commented on test/test_profile_source_info.py:
```diff
@@ -0,0 +1,77 @@
+        num_events_populated = num_events_populated_with_source_debug(prof.function_events)
+        self.assertEqual(
+                num_events_populated,
+                90,
```
this is a bad test - number of events might changes if we refactor ATen internals and organize functions differently.

kimishpatel replied:
Sure. Will fix it.

dzhulgakov reviewed (changes requested):
idea is nice, see some comment on implementation

@zdevito or @Krovatkin or @ZolotukhinM are the right folks to review it. It'd be nice to also think how it can fit together with debug symbols for other needs - we'd need something for fusion/backends as well

kimishpatel commented:
For some reason I am not able to add any reviewers here. Is that something I am missing?

ZolotukhinM commented:
> For some reason I am not able to add any reviewers here. Is that something I am missing?

You need to click the gear icon near "Reviewers" to add more people.

dzhulgakov reviewed (commented):
I'd wait for @zdevito @jamesr66a comments. But the interface between autograd profiler and jit should be changed imho - it's too fragile.

One option - is to just make JIT create its own autograd profiler scopes for each instruction (they are pretty cheap and can be nested)

kimishpatel committed:
Populate profiling events with source info. (#21668)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/21668

This changelist adds:
1) Information about where the operators are in the model source.
2) Populates this information in chrome trace.
3) Enables use of python interface for profiler to get source debug info.

Differential Revision: D15772906

fbshipit-source-id: 2759d33b7e068763ac312dd8d79f939b4b500d2d

zdevito reviewed (commented):
I am not comfortable exporting information about interpreter instructions out of the interpreter. The implementation is subject to change, and having this infrastructure makes doing that change more difficult. I already re-wrote the interpreter almost entirely this month and if these changes were in place it would have been much more difficult to do so. 

I think there is a much simpler approach that would get pretty equivalent information. Implement one function like this:

```
thread_local Node* currently_executing_node = nullptr
c10::optional<SourceRange> findCurrentlyExecutingRange() {
  if (!currently_execuing_node)
   return nullopt;
  return currently_executing_node->sourceRange();
}
```
In interpreter:
```
currently_executing_node = af.instructions_source_[af.pc]
switch(...)
```

This reduces the coupling since interpreter does not need to know about the profiler. In the future, we can extend `findCurrentlyExecutingRange` to extract a more full stack trace.

kimishpatel commented:
@zdevito, ok agreed in principle, but if we gonna store sourceRange as part of the events then we will be copying shared_ptr<Source> right? Doesn't this make profiling slow? Is that an acceptable cost?

jamesr66a reviewed (commented):
From offline discussion I think you're rethinking the approach here, but I left some small comments in case this ends up being the selected approach after all

kimishpatel commented on test/cpp/jit/test.cpp:
```diff
@@ -39,54 +40,56 @@ using namespace torch::jit::test;
+  _(ScriptObject)                         \
+  _(SaveExtraFilesHook)                   \
+  _(ProfileSourceDebugInfoSingleThread)   \
+  _(ProfileSourceDebugInfoMultiThreaded)
```
Actually I did the formatting manually here.

kimishpatel commented:
@jamesr66a, regarding using compilation mutex, this has the same issue as the other Zach mentioned regarding implementation leakage. So I think it would be better to keep it here. This might change anyway for better design that does not leak and rely on interpreter's implementation details.

kimishpatel commented:
@zdevito, Are you concern is that PC is being stored as part of event? Or is it that stringified instructions are stored, by this PR, in a std::vector and PC is used to look them up? If latter we can store it in map/unordered_map. Let me know. Thanks!

zdevito commented:
Both "instructions" and "pc" are are interpreter details. I am concerned that those leak from the interpreter. It is fine to expose the executing "Node*" and the source range because these are not internal details of the interpreter.  I haven't reviewed the PR for potential perf issues yet.

kimishpatel commented:
Thanks @zdevito. The perf issue I was concerned about what storing SourceRange obtained via Node*. SourceRange has `std::shared_ptr<Source> source_`. This, I assume, has perf implication due to atomic updates on shared ptr. I understand your concerns though. I dont have a clear solution yet that avoids the issue I mentioned and addresses your concern too.
